src/node_messages.cc

```diff
@@ -0,0 +1,65 @@
+
+#include "node_messages.h"
+#include "node.h"
+#include "v8.h"
+#include "util.h"
+#include "env.h"
+#include "env-inl.h"
+
+#define TYPE_ERROR(msg) env->ThrowTypeError(msg)
+
+// Used to be a macro, hence the uppercase name.
+#define DEFINE_MSG_CONSTANT(target, constant)                                 \
+  do {                                                                        \
+    v8::Isolate* isolate = target->GetIsolate();                              \
+    v8::Local<v8::String> constant_name =                                     \
+        v8::String::NewFromUtf8(isolate, #constant);                          \
+    v8::Local<v8::Number> constant_value =                                    \
+        v8::Number::New(isolate, static_cast<double>(MSG_ ## constant));      \
+    v8::PropertyAttribute constant_attributes =                               \
+        static_cast<v8::PropertyAttribute>(v8::ReadOnly | v8::DontDelete);    \
+    (target)->ForceSet(constant_name, constant_value, constant_attributes);   \
+  }                                                                           \
+  while (0)
+
+namespace node {
+
+using v8::Local;
+using v8::Object;
+using v8::String;
+using v8::Value;
+using v8::FunctionCallbackInfo;
+
+static void NodeMsg(const FunctionCallbackInfo<Value>& args) {
```

Comment:
part of the goal is to make it as simple as possible to translate the messages generated by both the native and js source all in one go... if at all possible, we should be able to bundle the messages into the single src/messages/{locale}/message.h file so that all we need to do to build a translated/localized node binary is swap that one file. Eventually, if we ever get to a point where ICU is on all the time, we can easily swap the messages.h out for an ICU resource bundle.

For the js messages, caching them into an object or array to eliminate having to do multiple runtime lookups would likely be the best optimization. In fact, we could easily prime that pump by having an API that returns the most commonly used / likely messages pre-cached.

This is quite definitely still a work in progress and I'm going to be benchmarking the heck out of this to make sure the impact is not too much of a burden.